kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala

```diff
@@ -319,4 +323,50 @@ private[kyuubi] class EngineRef(
       }
     }
   }
+
+  def getAdaptivePoolId(poolSize: Int): Int = {
+    val sessionThreshold = conf.get(ENGINE_POOL_ADAPTIVE_SESSION_THRESHOLD)
+    val metricsSpace =
+      s"/metrics/${serverSpace}_${KYUUBI_VERSION}_${shareLevel}_${engineType}/$user"
+    DiscoveryClientProvider.withDiscoveryClient(conf) { client =>
+      tryWithLock(client) {
+        if (client.pathNonExists(metricsSpace)) {
+          client.create(metricsSpace, "PERSISTENT")
+        }
+      }
+      val metrics = client.getChildren(metricsSpace)
+      if (metrics.isEmpty) {
+        return Random.nextInt(poolSize)
+      } else {
+        engineType match {
+          case SPARK_SQL =>
+            val engineMetricsMap = metrics.map(p =>
+              new String(client.getData(s"$metricsSpace/$p"))
+                .split(";")
+                .map(_.split("=", 2))
+                .filter(_.length == 2)
+                .map(kv => (kv.head, kv.last.toInt))
+                .toMap)
+            if (engineMetricsMap.isEmpty) {
+              return Random.nextInt(poolSize)
+            } else {
+              val sortedEngineMetrics = engineMetricsMap.sortBy { map =>
+                (
+                  map.getOrElse("openSessionCount", sessionThreshold),
+                  map.getOrElse("activeTask", 0))
+              }
+              val candidate = sortedEngineMetrics.head
```

Comment:
Why pick the head without looping the candidates here?

Reply:
We should choose the lightest load engine